engine/access/state_stream/backend/handler.go

```diff
@@ -134,71 +154,195 @@ func (h *Handler) SubscribeEvents(request *executiondata.SubscribeEventsRequest,
-			return nil
-		}
-
-		resp, ok := v.(*EventsResponse)
-		if !ok {
-			return status.Errorf(codes.Internal, "unexpected response type: %T", v)
-		}
 
+	return func(resp *SubscribeEventsResponse) error {
 		// check if there are any events in the response. if not, do not send a message unless the last
 		// response was more than HeartbeatInterval blocks ago
 		if len(resp.Events) == 0 {
 			blocksSinceLastMessage++
 			if blocksSinceLastMessage < heartbeatInterval {
-				continue
+				return nil
 			}
 			blocksSinceLastMessage = 0
 		}
 
 		// BlockExecutionData contains CCF encoded events, and the Access API returns JSON-CDC events.
 		// convert event payload formats.
 		// This is a temporary solution until the Access API supports specifying the encoding in the request
-		events, err := convert.EventsToMessagesWithEncodingConversion(resp.Events, entities.EventEncodingVersion_CCF_V0, request.GetEventEncodingVersion())
+		events, err := convert.EventsToMessagesWithEncodingConversion(resp.Events, entities.EventEncodingVersion_CCF_V0, eventEncodingVersion)
 		if err != nil {
 			return status.Errorf(codes.Internal, "could not convert events to entity: %v", err)
 		}
 
-		err = stream.Send(&executiondata.SubscribeEventsResponse{
-			BlockHeight: resp.Height,
-			BlockId:     convert.IdentifierToMessage(resp.BlockID),
-			Events:      events,
+		err = send(&executiondata.SubscribeEventsResponse{
+			BlockHeight:    resp.Height,
+			BlockId:        convert.IdentifierToMessage(resp.BlockID),
+			Events:         events,
+			BlockTimestamp: timestamppb.New(resp.BlockTimestamp),
+			MessageIndex:   resp.MessageIndex,
 		})
 		if err != nil {
 			return rpc.ConvertError(err, "could not send response", codes.Internal)
 		}
+
+		return nil
+	}
+}
+
+// getEventFilter returns an event filter based on the provided event filter configuration.// If the event filter is nil, it returns an empty filter.
+// Otherwise, it initializes a new event filter using the provided filter parameters,
+// including the event type, address, and contract. It then validates the filter configuration
+// and returns the constructed event filter or an error if the filter configuration is invalid.
+// The event filter is used for subscription to events.
+//
+// Parameters:
+// - eventFilter: executiondata.EventFilter object containing filter parameters.
+//
+// Expected errors during normal operation:
+// - codes.InvalidArgument - if the provided event filter is invalid.
+func (h *Handler) getEventFilter(eventFilter *executiondata.EventFilter) (state_stream.EventFilter, error) {
+	filter := state_stream.EventFilter{}
+	if eventFilter != nil {
+		var err error
+		filter, err = state_stream.NewEventFilter(
+			h.eventFilterConfig,
+			h.chain,
+			eventFilter.GetEventType(),
+			eventFilter.GetAddress(),
+			eventFilter.GetContract(),
+		)
+		if err != nil {
+			return filter, status.Errorf(codes.InvalidArgument, "invalid event filter: %v", err)
+		}
 	}
+	return filter, nil
 }
```

Comment:
let's flatten this. it makes the code easier to read
```suggestion
func (h *Handler) getEventFilter(eventFilter *executiondata.EventFilter) (state_stream.EventFilter, error) {
	if eventFilter == nil {
		return nil, nil
	}

	filter, err := state_stream.NewEventFilter(
		h.eventFilterConfig,
		h.chain,
		eventFilter.GetEventType(),
		eventFilter.GetAddress(),
		eventFilter.GetContract(),
	)
	if err != nil {
		return filter, status.Errorf(codes.InvalidArgument, "invalid event filter: %v", err)
	}

	return filter, nil
}
```

Reply:
[6d2d01d](https://github.com/onflow/flow-go/pull/5602/commits/6d2d01db147e50d00e9f19eb18f876e981707444)